File: packages/rust/tests/fixtures/unit_isolation/python/red/widget_test.py
# Red fixture for Python unit isolation (#42, `unmocked-collaborator`). A unit test
# must isolate the unit under test, so importing a first-party collaborator without
# mocking it is the violation — patch it (by string, in a fixture) instead.
from myproject.widget import build  # unit under test — never a collaborator

from myproject.ledger import record  # first-party collaborator, NOT mocked → violation


def describe_build():
    def it_builds():
        # Uses the real `record` collaborator — not isolated.
        assert build(record) is not None

File: packages/rust/tests/fixtures/unit_isolation/typescript/red/widget.test.ts
import { describe, expect, it, vi } from 'vitest';

import { makeWidget } from './widget'; // unit under test — never a collaborator
import { format } from './formatter'; // first-party collaborator, NOT mocked → violation
import { chunk } from 'lodash'; // external collaborator, NOT mocked → violation
import { log } from './logger'; // a collaborator, but mocked below → fine

vi.mock('./logger');

describe('makeWidget', () => {
  it('builds a widget', () => {
    const widget = makeWidget({ format, chunk, log });
    expect(widget).toBeTruthy();
  });
});
